mechanicalsoup/stateful_browser.py

```diff
@@ -260,23 +276,49 @@ def follow_link(self, link=None, *args, **kwargs):
 
         :return: Forwarded from :func:`open_relative`.
         """
-        if not hasattr(link, 'attrs') or 'href' not in link.attrs:
-            try:
-                link = self.find_link(link, *args, **kwargs)
-            except LinkNotFoundError:
-                if self.get_debug():
-                    print('follow_link failed for', link)
-                    self.list_links()
-                    self.launch_browser()
-                raise
-
+        link = self.find_link(link, *args, **kwargs)
         referer = self.get_url()
         request_kwargs = dict()
         if referer is not None:
             request_kwargs['headers'] = {'Referer': referer}
 
         return self.open_relative(link['href'], **request_kwargs)
 
+    def download_link(self, filename, link=None, *args, **kwargs):
+        """Downloads the contents of a link to a file. This function behaves
+        similarly to :func:`follow_link`, but the browser state will
+        not change when calling this function.
+
+        :param filename: Filesystem path where the page contents will be
+            downloaded. If the file already exists, it will be overwritten.
+        :param link: If this parameter has an *href*-attribute,
+            download the page pointed-to by this attribute. Otherwise,
+            consider this as the first argument forwarded
+            to :func:`find_link`.
+        :param \*args&\*\*kwargs: Arguments forwarded
```

Comment:
The description of these arguments is a bit obsolete with the addition of the `link` param. Perhaps we can just leave it as "Additional arguments forwarded to find_link"?